query-planner-js/src/Scope.ts

```diff
@@ -0,0 +1,259 @@
+import { DirectiveNode, GraphQLCompositeType, GraphQLObjectType, isTypeSubTypeOf, ValueNode } from "graphql";
+import { QueryPlanningContext } from "./QueryPlanningContext";
+
+/**
+ * A chain of type conditions leading to a particular point within an object.
+ *
+ * For instance, for a query like:
+ * ```
+ * {
+ *   v {
+ *     f1
+ *     ... on X {
+ *       f2
+ *       ... on Y @myDirective {
+ *         f3
+ *       }
+ *     }
+ *   }
+ * }
+ * ```
+ * and assuming the type of field `v` is `V`, then the scope for `f1` would be just type `V`,
+ * while the one for `f2` is the chain `X <- V` and the one for `f3` is the chain `Y <- X <- V`.
+ *
+ * Mainly, the scope contains enough information to 1) to compute the set of possible runtime
+ * type an object can have at a given point in the query (@see `possibleRuntimeTypes`) and
+ * 2) recreate minimal type conditions to obtain that set of possible runtimes.
+ *
+ * Additionally, the scope keep track of any directive that is applied on one of the type
+ * condition it represents. So in the example below, the scope will aslo track that for
+ * reaching `f3` there is a `@myDirective` directive on the `Y` condition.
+ *
+ * Do note that scope attempts to be "minimal" in that they eschew unecessary conditions. For
+ * example, in query:
+ * ```
+ * {
+ *   v {
+ *     ... on X {
+ *       ... on V {
+ *         f
+ *       }
+ *     }
+ *   }
+ * }
+ * ```
+ * still assuming that `v` is of type `V` and assuming that X is a subtype of V, the scope for field
+ * `f` will simply be `V <- X` (instead of `V <- X <- V`) as the outermost `... on V` condition adds
+ * nothing (but scope always preserve the most inner type condition as for a field, this is the type
+ * in which the field is defined). Additionally, if type `X` happens to a _super type_ of `V`, the
+ * scope for `f` would simply be `V`.
+ */
+export class Scope {
+  private cachedRuntimeTypes?: ReadonlyArray<GraphQLObjectType>;
+  private cachedIdentityKey: string | undefined = undefined;
+
+  private constructor(
+    private readonly context: QueryPlanningContext,
+
+    /** The innermost type of this scope. */
+    public readonly parentType: GraphQLCompositeType,
+    /** Directives associated to `parentType`. */
+    public readonly directives?: ReadonlyArray<DirectiveNode>,
+    /** Then enclosing scope, that is the next scope in the chain. */
+    public readonly enclosing?: Scope
+  ) {
+  }
+
+  /**
+   * Creates a new scope given a top-level type.
+   *
+   * @param context - the query planning context in which this scope is created.
+   * @param parentType - the top-level (and only) type of the created scope.
+   * @returns the newly created scope.
+   */
+  static create(context: QueryPlanningContext, parentType: GraphQLCompositeType) {
```

Comment:
The reason is to not expose the constructor because I don't want the 2 last arguments to be provided externally from the class, I want to force the use of `refine`.

Note that I reckon that, as always with programming, this could be written some other way. For instance, we could have only the constructor public (and neither `create` nor `refine`) and have the constructor handle the `refine` logic`. But I find this less clean in its usage. I like expressing the construction of scope as the combination of 1) creating a "root" scope, or 2) refining another scope (applying a type condition really).

Reply:
Agreed that the use of `refine` shows the intention a lot better. My first thought was that it should be possible to write a public overload declaration for the constructor, but unfortunately TypeScript doesn't allow mixing access levels in overloads. So I think a factory method is a fine alternative. Again, this seems like something we may want to write down (including a convention to use `create` as a default factory method).